### PR TITLE
Fix endless range include

### PR DESCRIPTION
**Pull request by aldhsu**
#### Problem 
See https://github.com/rails/rails/issues/36451
Core extension on ranges calls `#last`.
Ruby 2.6.0 Endless ranges blow up on calling `#last`:
```
ruby -e "(1..).last"                                                                                                                                                                               2.6.2 help-plan-scoring b2c08bc ✗
Traceback (most recent call last):
        1: from -e:1:in `<main>'
-e:1:in `last': cannot get the last element of endless range (RangeError)
```

#### Solution
-`#end` seems directly substitutable, (`#last` acts like `#end` if `#last` is called without args) and does not blow up and returns nil if no end is defined, from the source of `#last`
```C
               static VALUE
range_last(int argc, VALUE *argv, VALUE range)
{
    if (NIL_P(RANGE_END(range))) {
        rb_raise(rb_eRangeError, "cannot get the last element of endless range");
    }
    if (argc == 0) return RANGE_END(range);
    return rb_ary_last(argc, argv, rb_Array(range));
}
```
Source of `#end`:
```C
               static VALUE
range_end(VALUE range)
{
    return RANGE_END(range);
}
```

### CODE REVIEW

File: activesupport/test/core_ext/range_ext_test.rb [hunk 1]
```diff
@@ -60,6 +60,10 @@ def test_should_include_other_with_exclusive_end
     assert((1..10).include?(1...11))
   end
 
+  def test_include_with_endless_range
+    assert((1..).include?(2))
```
Comment (aldhsu) [inline activesupport/test/core_ext/range_ext_test.rb]: Not sure how to handle the CI failures for this test with Ruby < `2.6.0`.

Comment (sikachu) [reply activesupport/test/core_ext/range_ext_test.rb]: You can wrap the whole thing around conditional:

```ruby
if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new("2.6.0")
  def test_include_with_endless_range
    # ...
```

**Comment by aldhsu** [comment]
Rebased on master.

File: activesupport/test/core_ext/range_ext_test.rb [hunk 2]
```diff
@@ -60,6 +60,12 @@ def test_should_include_other_with_exclusive_end
     assert((1..10).include?(1...11))
   end
 
+  if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new("2.6.0")
+    def test_include_with_endless_range
+      assert((1..).include?(2))
```
Comment (eugeneius) [inline activesupport/test/core_ext/range_ext_test.rb]: To prevent this test from breaking CI on Rails 2.5, you can use `eval` to defer parsing until runtime:

```suggestion
      assert(eval("1..").include?(2))
```

🙇  @sikachu for coming up with this solution! https://github.com/rails/rails/pull/34906#discussion_r247001824

**Comment by eugeneius** [comment]
Could you add a changelog entry and squash your commits? Otherwise, this looks good to me. 👍

**Review by eugeneius — APPROVED** [review]

File: activesupport/lib/active_support/core_ext/range/include_time_with_zone.rb [hunk 3]
```diff
@@ -11,7 +11,7 @@ module IncludeTimeWithZone #:nodoc:
     def include?(value)
       if first.is_a?(TimeWithZone)
```
Comment (kamipo) [inline activesupport/lib/active_support/core_ext/range/include_time_with_zone.rb]: Can you address this line as well for beginless ranges?

```
irb(main):001:0> RUBY_VERSION
=> "2.7.0"
irb(main):002:0> a = (..1)
=> ..1
irb(main):003:0> a.first
Traceback (most recent call last):
        6: from /Users/kamipo/.rbenv/versions/2.7.0-dev/bin/irb:23:in `<main>'
        5: from /Users/kamipo/.rbenv/versions/2.7.0-dev/bin/irb:23:in `load'
        4: from /Users/kamipo/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/irb-1.1.0.pre.2/exe/irb:11:in `<top (required)>'
        3: from (irb):3
        2: from (irb):4:in `rescue in irb_binding'
        1: from (irb):4:in `first'
RangeError (cannot get the first element of beginless range)
```

File: activesupport/test/core_ext/range_ext_test.rb [hunk 4]
```diff
@@ -60,6 +60,16 @@ def test_should_include_other_with_exclusive_end
     assert((1..10).include?(1...11))
   end
 
+  if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new("2.6.0")
+    def test_include_with_endless_range
+      assert(eval("1..").include?(2))
+    end
+
+    def test_include_with_beginless_range
```
Comment (eugeneius) [inline activesupport/test/core_ext/range_ext_test.rb]: Beginless ranges are an upcoming feature in Ruby 2.7; this test will needs its own version guard.

File: activesupport/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,7 @@
+*   Fix `Range#.include?` to work with Ruby 2.6 beginless and endless ranges.
```
Comment (eugeneius) [inline activesupport/CHANGELOG.md]:
```suggestion
*   Fix `Range#include?` to work with beginless and endless ranges.
```

Including both Ruby 2.6 and Ruby 2.7 would make the sentence too hard to parse, and I don't think mentioning the versions that introduce the features is that important anyway.

**Comment by aldhsu** [comment]
Thanks @kamipo and @eugeneius for the reviews!

**Comment by pond** [comment]
This misses some other cases - see https://github.com/rails/rails/issues/37172 and comments therein. I'll take the approach established by this PR for tests etc., & get a patch together to fill in the rest of the gaps.
